### DISCUSSION TITLE
Some questions about the crazyflie-firmware

### DISCUSSION THREAD

**Discussion by 7enHom** [bitcraze/discussions#1280] [Q&A] [answered]
Hi！
I'm doing some work with crazyflie2.1 and this needs to be done by changing the firmware, and so far I'm having some problems.

1. I would like to know if the concept of time exists in crazyflie-firmware when it is running. For example, if I send a command to crazyflie via commander.send_setpoint(), can I log the time when this command was run in the firmware and then transfer this information by adding a logging group？

2. As far as I know, crazyflie's roll angle rotation range is [-180, 180] degrees, which means that if I flip the crazyflie 360 degrees, its roll angle will change from 0 to 180 degrees, and then suddenly change from -180 degrees to 0 degrees, which is a discontinuous process. I'm wondering if it's possible to change this range in the firmware so that the range turn to [0, 360] degrees.

Thank you very much for answering my questions!

**Comment by knmcguire** [comment]
Hi! Let me try to answer your questions.

1- Yes we have! you'll need to look for the usec_time.c, with usecTimestamp which gives the time in microseconds: https://github.com/bitcraze/crazyflie-firmware/blob/9bcb278a03f1e659e4a42f28a91a43def8098757/src/hal/src/usec_time.c#L86

2- That might be tricky and require dinging into the state estimation processes. Perhaps it would be good to first think if this is something you want. Flipping 360 degrees with [0 360] is a discontinuous process as well with the difference of the sigsaw log being in the positive figure realm only. At least with the [-180 180], it is clear where the normal zero degrees is, and which ways are regarded the negative and positive rotation.

**Comment by 7enHom** [comment]
Thanks a lot！

For questions 1, I think the usecTimestamp() should be used like this but I haven't had a time to confirm it:

t1 = usecTimestamp();
if (RATE_DO_EXECUTE(ATTITUDE_RATE, tick)) {
    ...;
    ...;
}
t2 = usecTimestamp();

Then t = t2 - t1 is probably the running time, is that right?


And for questions 2, I think what I want is that if the flip range is [0 360] degrees, then I can keep a good record of the number of turns the crazyflie has flipped, e.g., count_flip += 1 or something like that.

**Reply by gemenerik** [reply, marked as answer]
1. Yes, I think something like this should work.
2. I'll second that changing the coordinate system in the entire firmware is way too complex for what you want to do.There is nothing withholding you from shifting the value by adding 180° to the roll angle in e.g. your app or script!

**Reply by 7enHom** [reply]
Thanks!
